format/apple/bplist/ns_keyed_archiver.jq

```diff
@@ -0,0 +1,77 @@
+def from_ns_keyed_archiver(root):
+  (  
+    . as 
+      {
+        "$objects": $objects
+      }
+  | def _f($id; $seen_ids):
+      def _r($id):
+        if $seen_ids | has("\($id)") then "cycle-\($id)"
+        else _f($id; $seen_ids | ."\($id)" = true)
+        end;
+      ( $objects[$id]
+      | type as $type |
+        if $type == "string" and . == "$null" then null
+        elif $type |
+          . == "number"
+          or . == "boolean"
+          or . == "null"
+          or . == "string" then .
+        elif $type == "array" then . # TODO: does this happen?
+        elif $type == "object" then
+          ( ."$class" as $class
+          | if $class == null then # TODO: what case is this?
+              with_entries(
+                if (.value | type == "object") 
+                    and (.value | has("cfuid")) 
+                    then .value |= _r(.cfuid) end
+              )
+            else
+              ( $objects[$class.cfuid]."$classname" as $cname
+              | if $cname == "NSDictionary"
+                  or $cname == "NSMutableDictionary" then
+                  # transform arrays [key_id1, key_id2,...] and [obj_id1, obj_id2,..] into {key: obj, ...}
+                  ( [."NS.keys", ."NS.objects"]
+                  | transpose
+                  | map({key: _r(.[0].cfuid), value: _r(.[1].cfuid)})
+                  | from_entries
+                  )
+                elif $cname == "NSArray"
+                  or $cname == "NSMutableArray"
+                  or $cname == "NSSet"
+                  or $cname == "NSMutableSet" then
+                  ( ."NS.objects"
+                  | map(_r(.cfuid))
+                  )
+                elif $cname == "NSData" or $cname == "NSMutableData" then ."NS.Data" # TODO: will be a json string?
+                elif $cname == "NSDate" then "NS.time"
+                elif $cname == "NSNull" then null
+                elif $cname == "NSAttributedString"
+                  or $cname == "NSMutableAttributedString" then
+                  _r(.NSString)
+                elif $cname == "NSUUID" then ."NS.uuidbytes" # TODO: will be a json string?
+                else
+                  # replace class ID with classname, and dereference all cfuid values.
+                  ."$class" = $cname |
+                  with_entries(
+                    if (.value | type == "object") 
+                      and (.value | has("cfuid")) 
+                      then .value |= _r(.cfuid) end
+                  )
+                end
+              )
+            end
+          )
+        end
+      );
+    def _f($id): _f($id; {"\($id)": true});
+    _f(root)
+  );
+
+def from_ns_keyed_archiver:
+  ( . as 
+      {
+        "$top": {root: $root}
+      }
+    | from_ns_keyed_archiver($root.cfuid)
```

Comment:
Maybe this should be `from_ns_keyed_archiver($root?.cfuid // 1)` as a sensible fallback option? Or is it better if it just fails?

Reply:
Are there files in the wild that seems assume root uid 1? could it be misleading if not and it works?

Reply:
Ah now i see. Now you have to write something like:
```
fq 'torepr | from_ns_keyed_archiver(."$top"."796BFF22-6712-4486-A32C-A1C5DB3273BA"."cfuid")'  ~/Downloads/com.apple.networkextension.plist
```

Reply:
`com.apple.networkextension.plist` is a good example of one that doesn't have a `$root` value and begins decoding at index 1, but I do lean towards forcing the user to specify an alternative root so that there aren't potential false positives and missing data in the resulting representation.

Reply:
And lookin for a UUID key might quite hacky? maybe best to default to 1 then?

Reply:
Ok, let's not default then 1 then. Maybe then should throw some more helpful error on null $root? and possibly add example to doc about uid 1?